rhp/v4/rhp.go

```diff
@@ -0,0 +1,768 @@
+	riskedCollateral := prices.Collateral.Mul64(fc.Filesize).Mul64(renewal.NewContract.ExpirationHeight - prices.TipHeight)
+
+	// total collateral includes the additional requested collateral and
+	// the risked collateral from the existing storage.
+	renewal.NewContract.TotalCollateral = rp.Collateral.Add(riskedCollateral)
+	// missed host value should only include the new collateral value
+	renewal.NewContract.MissedHostValue = rp.Collateral
+
+	// storage cost is the difference between the new and old contract since the
+	// old contract already paid for the storage up to the current expiration
+	// height.
+	storageCost := prices.StoragePrice.Mul64(fc.Filesize).Mul64(renewal.NewContract.ExpirationHeight - fc.ExpirationHeight)
+
+	// host output value includes the locked + risked collateral, the additional
+	// storage cost, and the contract price.
+	renewal.NewContract.HostOutput.Value = renewal.NewContract.TotalCollateral.Add(storageCost).Add(prices.ContractPrice)
+
+	// if the existing locked collateral is greater than the new required
+	// collateral, the host will only lock the new required collateral. Otherwise,
+	// roll over the existing locked collateral. The host will need to fund
+	// the difference.
+	if fc.TotalCollateral.Cmp(renewal.NewContract.TotalCollateral) > 0 {
+		renewal.HostRollover = renewal.NewContract.TotalCollateral
+	} else {
+		renewal.HostRollover = fc.TotalCollateral
+	}
+
+	// if the remaining renter output is greater than the required allowance,
+	// only roll over the new allowance. Otherwise, roll over the remaining
+	// allowance. The renter will need to fund the difference.
+	if fc.RenterOutput.Value.Cmp(rp.Allowance) > 0 {
+		renewal.RenterRollover = rp.Allowance
+	} else {
+		renewal.RenterRollover = fc.RenterOutput.Value
+	}
+	return renewal, Usage{
+		RPC:              prices.ContractPrice,
+		Storage:          renewal.NewContract.HostOutput.Value.Sub(renewal.NewContract.TotalCollateral).Sub(prices.ContractPrice),
+		RiskedCollateral: renewal.NewContract.TotalCollateral.Sub(renewal.NewContract.MissedHostValue),
+	}
+}
+
+// RefreshContract creates a contract renewal for the refresh RPC.
+func RefreshContract(fc types.V2FileContract, prices HostPrices, rp RPCRefreshContractParams) (types.V2FileContractRenewal, Usage) {
+	var renewal types.V2FileContractRenewal
+
+	// clear the old contract
+	renewal.FinalRevision = fc
+	renewal.FinalRevision.RevisionNumber = types.MaxRevisionNumber
+	renewal.FinalRevision.RenterSignature = types.Signature{}
+	renewal.FinalRevision.HostSignature = types.Signature{}
+
+	// create the new contract
+	renewal.NewContract = fc
+	renewal.NewContract.RevisionNumber = 0
+	renewal.NewContract.RenterSignature = types.Signature{}
+	renewal.NewContract.HostSignature = types.Signature{}
+	// add the additional allowance and collateral
+	renewal.NewContract.RenterOutput.Value = fc.RenterOutput.Value.Add(rp.Allowance)
```

Comment:
I'm not sure adding `Additional` is significantly better since a refreshed or renewed contract will both have at least `allowance` and `collateral` available to use.